src/backend/columnar/columnar_tableam.c

```diff
@@ -585,18 +547,37 @@ columnar_index_fetch_tuple(struct IndexFetchTableData *sscan,
 	}
 	else if (StripeWriteInProgress(stripeMetadata))
 	{
-		/* similar to aborted writes .. */
-		Assert(snapshot->snapshot_type == SNAPSHOT_DIRTY);
+		if (stripeMetadata->visibleToSelf)
+		{
+			/*
+			 * Stripe is visible to me, so it must be written by me. Since
+			 * caller might want to use tupleslot datums for some reason, do
+			 * another look-up, but this time by first flushing our writes.
+			 */
+			ColumnarReadFlushPendingWrites(scan->cs_readState);
 
-		/*
-		 * Stripe that "might" contain the tuple with rowNumber is not
-		 * flushed yet. Here we set all attributes of given tupleslot to NULL
-		 * before returning true and expect the indexAM callback that called
-		 * us --possibly to check against constraint violation-- blocks until
-		 * writer transaction commits or aborts, without requiring us to fill
-		 * the tupleslot properly.
-		 */
-		memset(slot->tts_isnull, true, slot->tts_nvalid);
+			/*
+			 * Fill the tupleslot and fall through to return true, it
+			 * certainly exists.
+			 */
+			ColumnarReadRowByRowNumber(scan->cs_readState, rowNumber,
+									   slot->tts_values, slot->tts_isnull);
+		}
+		else
+		{
+			/* similar to aborted writes .. */
+			Assert(snapshot->snapshot_type == SNAPSHOT_DIRTY);
+
+			/*
+			 * Stripe that "might" contain the tuple with rowNumber is not
+			 * flushed yet. Here we set all attributes of given tupleslot to NULL
+			 * before returning true and expect the indexAM callback that called
+			 * us --possibly to check against constraint violation-- blocks until
+			 * writer transaction commits or aborts, without requiring us to fill
+			 * the tupleslot properly.
+			 */
+			memset(slot->tts_isnull, true, slot->tts_nvalid);
```

Comment:
Test with two sessions with an exclusion constraint, make second one blocked on first session, then commit the first session, see if error message is proper